jax/_src/lax/control_flow/common.py

```diff
@@ -62,9 +62,9 @@ def _initial_style_jaxpr(fun: Callable, in_tree, in_avals,
   closed_jaxpr = core.ClosedJaxpr(pe.convert_constvars_jaxpr(jaxpr), ())
   return closed_jaxpr, consts, out_tree
 
-@cache()
+@weakref_lru_cache
```

Comment:
I don't think this does what we want; we want the cache keys to hold weak refs to each of the `funs`, and for the entry to disappear when one of the components of `funs` has its last strong reference dropped.

As written, we're just holding a weakref to the first of `funs` (see [this comment](https://github.com/google/jax/blob/8dc1dff61098467dbd52b72eb542f2e205543c59/jax/_src/util.py#L266)).

Or maybe I'm mistaken?

Reply:
Oh, I see - I thought it used weak refs to all arguments. Thanks - it looks like there's not any good way to use `weakref_lru_cache` here without digging into the C implementation.

Reply:
Sorry, I didn't articulate it very well in the other comment!